File: 프로그래머스/lv1/김미래/가장 많이 받은 선물.cpp
#include <string>
#include <vector>
#include <sstream>
#include <map>
using namespace std;

int solution(vector<string> friends, vector<string> gifts) {
    int answer = 0;
    vector<vector<int>> history(friends.size(), vector<int> (friends.size(), 0));
    vector<int> degree(friends.size(), 0);
    map<string, int> index;
    
    for(int i=0; i<friends.size(); i++){
        index[friends[i]] = i;
    }
    
    for(int j=0; j<gifts.size(); j++){
        istringstream str(gifts[j]);
        string send, recieve;
        str >> send >> recieve;
        
        history[index[send]][index[recieve]] += 1;
    }
    
    for (int i = 0; i < history.size(); i++) {
        for (int j = 0; j < history.size(); j++) {
            degree[i] += history[i][j]; 
            degree[i] -= history[j][i];  
        }
    }

    
    for(int i=0; i< history.size(); i++){
        int now = 0;
        for(int j=0; j<history.size(); j++){
            if(i == j) continue;
            
            if(history[i][j] > history[j][i]){
                now ++;
            } else if(history[i][j]==history[j][i]){
                if(degree[i]>degree[j]) now++;
            }
        }
         if (now>answer)
        {
            answer=now;
        }
    }

    return answer;
}

File: 프로그래머스/lv1/이재표/가장 많이 받은 선물.java
import java.util.*;
class Solution {
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        Map<Integer,Integer>presentScore=new HashMap<>();
        Map<String,Integer>friendNum=new HashMap<>();
        int n=friends.length;
        int[][]present=new int[n][n];
        int max=-1;
        for(int i=0;i<friends.length;i++){
            friendNum.put(friends[i],i);
        }
        for(int i=0;i<gifts.length;i++){
            String[]gift=gifts[i].split(" ");
            int give=friendNum.get(gift[0]);
            int receive=friendNum.get(gift[1]);
            present[give][receive]+=1;
        }
        for(int i=0;i<n;i++){
            int score=0;
            for(int j=0;j<n;j++){
                score+=present[i][j];
            }
            for(int j=0;j<n;j++){
                score-=present[j][i];
            }
            presentScore.put(i,score);
        }
        
        
        
        for(int i=0;i<n;i++){
            int willrecieve=0;
            for(int j=0;j<n;j++){
                if(i==j){
                    continue;
                }
                int give=present[i][j];
                int recieve=present[j][i];
                if(give>recieve){
                    willrecieve+=1;
                }else if(give==recieve){
                    int givePresentScore=presentScore.get(i);
                    int recievePresentScore=presentScore.get(j);
                    if(givePresentScore>recievePresentScore){
                        willrecieve+=1;
                    }
                }
            }
            max=Math.max(max,willrecieve);
        }
        return max;
    }
}
